community/kernel/src/main/java/org/neo4j/kernel/impl/api/state/TxState.java

```diff
@@ -1014,6 +1015,33 @@ public PrimitiveLongReadableDiffSets indexUpdatesForScan( SchemaIndexDescriptor
         return diffs;
     }
 
+    @Override
+    public PrimitiveLongReadableDiffSets indexUpdatesForSuffixOrContains( SchemaIndexDescriptor descriptor, IndexQuery query )
+    {
+        descriptor.schema().getPropertyId(); // assert single property index
+
+        if ( indexUpdates == null )
+        {
+            return PrimitiveLongReadableDiffSets.EMPTY;
+        }
+        Map<ValueTuple, PrimitiveLongDiffSets> updates = indexUpdates.get( descriptor.schema() );
+        if ( updates == null )
+        {
+            return PrimitiveLongReadableDiffSets.EMPTY;
+        }
+        PrimitiveLongDiffSets diffs = new PrimitiveLongDiffSets();
+        for ( Map.Entry<ValueTuple,PrimitiveLongDiffSets> entry : updates.entrySet() )
+        {
+            if ( query.acceptsValue( entry.getKey().getOnlyValue() ) )
```

Comment:
why is this check only part of this retrieval of index updates?

Reply:
For `seek`, `prefix`, `range` and `scan`, these methods can query the `updates` map (a `HashMap` or `TreeMap`) for updates which exactly match the query - meaning that no filtering is needed.

For `suffix` and `contains` that is not possible though, so we need to scan all updates and filter for keys which actually match the query. This filtering was not done previously, and that bug is what we fixed.